### PR TITLE
events design

### PR DESCRIPTION
**Pull request by tudor-malene**
### Why is this change needed?

we need to implement events

### What changes were made as part of this PR:

design for events

### What are the key areas to look at

- ...

### :rotating_light: Definition of Done :rotating_light:
- [ ] Deployed into dev-testnet 
- [ ] Passes tests on dev-testnet
- [ ] [Changelog.md](https://github.com/obscuronet/go-obscuro/blob/main/docs/testnet/changelog.md) updated

### CODE REVIEW

File: design/Events_design.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+```
+
+A web app might request:
+- stream all "transfer" events where the 'to' field is my address
+- stream all "transfer" events from the USDC contract
+
+
+### How events work 
+
+The query made on the UI side is transformed in a server-side query on the node, and there is some logic after a tx is executed and events are emmitted, to match them against the filters requested by users and distribute them to the requester.(? - need to confirm this is 100% how it works) .
```
Comment (joeldudley) [inline design/Events_design.md:37]: Can you clarify in the doc whether the events are discarded after being distributed, or are stored for future subscribers? e.g. Can I connect to an Ethereum node and use a subscription to get a list of all events going back to the first block?

File: design/Events_design.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+3. The enclave runs all the usual filtering logic on this event, and determines if there is any subscription that requested it.
+
+4. Then, there is an extra step (inside the enclave as well) to determine whether that party is allowed to view that event. 
+
+This can be done by running an intersection between:
+  - all accounts found in the event topics plus the `from` of the transaction: `E1->(A1,A2,A3)` 
+  - and the accounts from the subscription. 
+
+If there is at least one element, the event is encrypted with the first of the result.
```
Comment (joeldudley) [inline design/Events_design.md:93]: "the event is encrypted with the first of the result" - can we reword this as ""the event is encrypted with the first address in the result set"? Found the original phrasing a bit confusing.

Comment (joeldudley) [inline design/Events_design.md:93]: "the event is encrypted with the first of the result" - I assume this is because the wallet extension has the keys to all the subscription accounts by definition, and so we don't need to encrypt with all the relevant keys, just one of them? If so, can you make it clear that's the reasoning in the doc? Think it might confuse the reader that we don't encrypt with _each_ matching key.

File: design/Events_design.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+```
+var subscription = web3.eth.subscribe('logs', {
+    address: '0x123456..', 
+    topics: ['0x12345...']
+}, function(error, result){
+    if (!error)
+        console.log(result);
+});
+- address - String|Array: An address or a list of addresses to only get logs from particular account(s).
+- topics - Array: An array of values which must each appear in the log entries. The order is important, if you want to leave topics out use null, e.g. [null, '0x00...']. You can also pass another array for each topic with options for that topic e.g. [null, ['option1', 'option2']]
```
Comment (joeldudley) [inline design/Events_design.md:27]: Can we add a bit to this doc explaining how topics are set? It's not clear to me, and I think it's necessary info for understanding the design. There seems to be some implication that a topic is automatically set based on the `from` and/or `to` fields, perhaps?

File: design/Events_design.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+For our RPC client it would an authenticated subscription.
+
+2. We need to pass into the enclave the "Subscription" objects, and create an rpc streaming for each. 
+
+3. The enclave runs all the usual filtering logic on this event, and determines if there is any subscription that requested it.
+
+4. Then, there is an extra step (inside the enclave as well) to determine whether that party is allowed to view that event. 
+
+This can be done by running an intersection between:
+  - all accounts found in the event topics plus the `from` of the transaction: `E1->(A1,A2,A3)`
```
Comment (joeldudley) [inline design/Events_design.md:90]: Can we clarify this to, "all accounts whose addresses are found in"? Took me a while to understand.

File: design/Events_design.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+
+#### Problem1: How do we authenticate event subscriptions? 
+We have a wallet extension with viewing key support, which we have to use ideally, to avoid creating another mechanism.
+
+#### Problem2: Assuming the enclave knows who is subscribed, how will the filtering work?
+
+E.g: Alice is requesting to listen to all transfer events. (this is a perfectly reasonable request on a public chain)
+
+There are 2 cases:
```
Comment (joeldudley) [inline design/Events_design.md:55]: I think we're ignoring an important case here: "The event was emitted by a smart contract as a result of executing a tx sent by Alice, but should not be seen by Alice.". In fact, there are really four cases:

- The event was emitted by a smart contract as a result of executing a tx sent by Alice, and should be seen by Alice
- The event was emitted by a smart contract as a result of executing a tx sent by Alice, but should not be seen by Alice
- The event was emitted by a smart contract as a result of executing a tx sent by Bob, and it should be seen by Alice
- The event was emitted by a smart contract as a result of executing a tx sent by Bob, and it should not be seen by Alice

I think in the rest of the doc, we assume that Alice always has a right to see events caused by transactions she submitted. But is that always true? I'm trying to think of an example - perhaps a winning bid in a gambling contract might emit a message destined for the contract owner _only_ telling them they need to replenish their prize pool?

Even if we decide that Alice gets to see any events caused by her transactions (e.g. because it's simpler), I think we should still acknowledge this as a limitation in the doc.

File: design/Events_design.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+This is possible because the subscription call is implemented on Obscuro, and made by the wallet_extension, so it doesn't have to be compatible with Ethereum.
+For our RPC client it would an authenticated subscription.
+
+2. We need to pass into the enclave the "Subscription" objects, and create an rpc streaming for each. 
+
+3. The enclave runs all the usual filtering logic on this event, and determines if there is any subscription that requested it.
+
+4. Then, there is an extra step (inside the enclave as well) to determine whether that party is allowed to view that event. 
+
+This can be done by running an intersection between:
```
Comment (joeldudley) [inline design/Events_design.md:89]: Don't know if you care, but I think if you indent these sentences you can cause them to be nested under their bullets to make things more readable (but maybe not?).

File: design/Events_design.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+  - all accounts found in the event topics plus the `from` of the transaction: `E1->(A1,A2,A3)` 
+  - and the accounts from the subscription. 
+
+If there is at least one element, the event is encrypted with the first of the result.
+
+6. For event `E2`, that has no addresses in the topics, there are a couple of options:
+
+    1. Because it is not address specific, we consider it a public event, and send it to everyone who requests it. Basically, we send it encrypted with the first viewing key from any subscription that matches it.
```
Comment (joeldudley) [inline design/Events_design.md:97]: "we send it encrypted" - why are we encrypting it if it's a public event?

File: design/Events_design.md (hunk 5 above)
Comment (tudor-malene) [reply design/Events_design.md:55]: umm. that's a very good point Joel.
wdyt, @CaisManai ?

File: design/Events_design.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+My personal preference is for Option 1. Which is in line with the other decisions around not changing the behaviour of public variables, and documenting it for developers.
+
+8. The encrypted event is bubbled out of the enclave on the right subscription channel, and then sent to the wallet extension.
+
+
+### Scenarios
+
+#### 1. Curious user deploys a contract that intends to leak Erc20 balances.
+This is prevented by the protections in place in the ERC20 contract, who will only allow the initiator of a tx to read the balance.
```
Comment (joeldudley) [inline design/Events_design.md:109]: I thought based on this doc that events were only emitted in response to transactions. But these would be calls, not transactions. Are events emittable for calls, etc.?

File: design/Events_design.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+
+#### 2. Curious user listens for all possible events.
+Under Option1. They will get back all events that don't have addresses.
```
Comment (joeldudley) [inline design/Events_design.md:113]: Maybe make it clear that by you're referring to Option 1 under point 6 above, and not to the first scenario here.

File: design/Events_design.md [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+`E1` is formed of multiple topics, 2 of which are addresses (`A2` and `A3`), and there are some other random values.
+
+An event `E2` is emitted by a transaction `TX2` from an address `A4`.
+`E2` is formed of multiple topics, none of which are addresses.
+
+
+### Solution
+
+1. We modify the subscription call to accept a list of owning accounts. 
+Each account must be signed with the latest viewing key, to prevent someone from asking random events, just to leak info.
```
Comment (joeldudley) [inline design/Events_design.md:78]: "Each account must be signed with the latest viewing key" - is this necessary? Can't the enclave just reject subscriptions where it does not have viewing keys for one or more of the accounts?

File: design/Events_design.md [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+An event `E1` is emitted after executing transaction `TX1` with a from address `A1`.
+`E1` is formed of multiple topics, 2 of which are addresses (`A2` and `A3`), and there are some other random values.
+
+An event `E2` is emitted by a transaction `TX2` from an address `A4`.
+`E2` is formed of multiple topics, none of which are addresses.
+
+
+### Solution
+
+1. We modify the subscription call to accept a list of owning accounts.
```
Comment (joeldudley) [inline design/Events_design.md:77]: We use "we" here in various places. Can you be explicit about what's done by the user, vs. what's done by the wallet extension? Hopefully this will help highlight what the impact on the user experience is (hopefully very little, it seems?).

File: design/Events_design.md (hunk 10 above)
Comment (joeldudley) [reply design/Events_design.md:78]: Can't you just reject the subscription if there is no vk for the account on the enclave side? Instead of insisting on another signature?

**Comment by joeldudley** [comment]
Do we need to discuss the behaviour on disconnect - both when the client terminates, and when the node terminates? How will a client app work out that it has been disconnected (vs there being no events)? How will the enclave detect that a client is no longer listening, and what clean-up will it do? Is there a mechanism to end a subscription?

I'm sure we'll probably borrow the Ethereum approach here, but it would be good to have a quick summary or link on how this works in Ethereum.

**Comment by BedrockSquirrel** [comment]
Great comments from Joel - v interested in the answer to some of them. I'll post the concern I raised in the discord just to keep it in the discussion: is this additional workload for an enclave going to be feasible at scale 

I'm not talking about long-term scaling issues, I'm worried that even say 20 actively browsing users that are using the same RPC endpoint might increase the enclave workload by multiples. Every time a user navigates around a dapp it might open and close subscriptions, each of which will be an RPC call from user -> host -> enclave to update the streaming filters.

Not supposed to pose problem without a half-baked solution...

Have we done many thought experiments around "what if all events are public". I think: 
- it would make privacy opt-in for contract developers
- it would require dapp developers to get creative on how they notify and interpret events without broadcasting too much

Pros:
- no complexity and full compatibility with eth tooling
- doesn't increase enclave workload

Cons:
- Devs porting apps with chatty events don't get privacy by default (private-versions have to be implemented, think this is true to some extent anyway with checks in methods)
- No idea if creative ideas to solve the problem are possible, like could trading on uniswap include a random ID and then watch for that to confirm the trade - no idea if that's feasible
- Maybe there are situations where it's mathematically impossible to notify what you're trying to notify without revealing info

File: design/Events_design.md [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+Each account must be signed with the latest viewing key, to prevent someone from asking random events, just to leak info. 
+The call will fail if there are no viewing keys for all those accounts.
+This is possible because the subscription call is implemented on Obscuro, and made by the wallet_extension, so it doesn't have to be compatible with Ethereum.
+For our RPC client it would an authenticated subscription.
+
+2. We need to pass into the enclave the "Subscription" objects, and create an rpc streaming for each. 
+
+3. The enclave runs all the usual filtering logic on this event, and determines if there is any subscription that requested it.
+
+4. Then, there is an extra step (inside the enclave as well) to determine whether that party is allowed to view that event.
```
Comment (joeldudley) [inline design/Events_design.md:87]: I think we should document somewhere that with this approach, we lose the ability to index public-event logs on addresses.

File: design/Events_design.md (hunk 4 above)
Comment (joeldudley) [inline design/Events_design.md:90]: Will it always be clear which topics are addresses vs not addresses? Is it possible some topics will be hashes (this seemed common based on the link you provided, Tudor) that will look address-like?

File: design/Events_design.md (hunk 5 above)
Comment (tudor-malene) [reply design/Events_design.md:55]: I think this is a moot point @joeldudley, because the events are returned in tx receipts as well, so the person who submitted a tx will always see those events.
I think we can safely make that assumption. I'll state it clearly in the design

**Comment by BedrockSquirrel** [comment]
Super clear and digestible now. Generally seems sensible to me, few thoughts:

- what's the lifespan for a subscription in the enclave, unsubscribe option + an expiry?

- still pretty worried about perf and DoS potential, but we can test for it and try to optimise/prioritise traffic etc. Maybe we need read-only nodes to services these requests in production (normal nodes but the host doesn't give it any transactions/other work).

- very minor information leak for the host about their users, host owner can see how many relevant events their subscribers are getting back. Tbf that's no different from it seeing how many transactions that user is submitting etc., just a measure of their activity I guess and not tied to their acc address at all

File: design/Events_design.md [hunk 13]
```diff
@@ -0,0 +1,300 @@
...
+
+### How events work
+
+`Note: this section might not be 100% accurate.`
+
+The query made on the UI side is transformed in a server-side query on the node, and there is some logic after a tx is
+executed and events are emmitted, to match them against the filters requested by users and distribute them to the
+requester.
+
+Todo: Can you clarify in the doc whether the events are discarded after being distributed, or are stored for future
```
Comment (joeldudley) [inline design/Events_design.md:64]: @tudor-malene I think you answer this above? "Apps can also request historic events starting from any block."

File: design/Events_design.md (hunk 10 above)
Comment (BedrockSquirrel) [reply design/Events_design.md]: @tudor-malene I guess WE could resubmit the original signed VK for UX? Weakens the security slightly from "requester has the acc private key" to "requester has got signed viewing key" but they should be no easier for a hacker to steal than the viewing keys themselves.

File: design/Events_design.md (hunk 13 above)
Comment (tudor-malene) [reply design/Events_design.md:64]: yeah , but not sure what happens to them. If they are stored in the db, or regenerated every time. Something to look into when implementing. Don't think it's material to the design though

File: design/Events_design.md [hunk 14]
```diff
@@ -0,0 +1,300 @@
...
+to implement the privacy concerns with as little disruption as possible.
+
+### Event visibility rules
+
+There are a couple of cases that must be considered in order to decide whether Alice is entitled to view an event:
+
+1. The event was emitted by a smart contract as a result of executing a Tx sent by Alice.
+2. The event was emitted as result of a Tx sent by Bob which is relevant to Alice. (See below for a definition of
+   relevancy.)
+3. The event was emitted as result of a Tx sent by Bob which is not relevant to Alice.
```
Comment (joeldudley) [inline design/Events_design.md:81]: The wording here is confusing IMO (not clear if we're talking about the event being relevant, or the transaction). Can we change it to "An event that is relevant to Alice was emitted as a result of a Tx sent by Bob" and "An event that is not relevant to Alice was emitted as a result of a Tx sent by Bob".

File: design/Events_design.md [hunk 15]
```diff
@@ -0,0 +1,300 @@
...
+
+*Note that we assume that all events emitted during a transaction signed by Alice, will be visible to Alice.*
```
Comment (joeldudley) [inline design/Events_design.md:83]: Can you clarify which of the three cases above we support? Based on this sentence, we support (1). Do we also support (2) and (3)?

File: design/Events_design.md [hunk 16]
```diff
@@ -0,0 +1,300 @@
...
+There are a couple of cases that must be considered in order to decide whether Alice is entitled to view an event:
+
+1. The event was emitted by a smart contract as a result of executing a Tx sent by Alice.
+2. The event was emitted as result of a Tx sent by Bob which is relevant to Alice. (See below for a definition of
+   relevancy.)
+3. The event was emitted as result of a Tx sent by Bob which is not relevant to Alice.
+
+*Note that we assume that all events emitted during a transaction signed by Alice, will be visible to Alice.*  
+
+The developer of a smart contract is responsible with making sure to not include any data that must be kept secret from
```
Comment (joeldudley) [inline design/Events_design.md:85]: Should this be "include any data in the event"? Wasn't clear to me.

File: design/Events_design.md (hunk 15 above)
Comment (joeldudley) [reply design/Events_design.md:83]: Ah, you cover this below. I think maybe get rid of this sentence here, since you cover below who gets to see what.

File: design/Events_design.md [hunk 17]
```diff
@@ -0,0 +1,300 @@
...
+
+1. An event is considered relevant to all account owners whose addresses are used as topics in the event.
+2. In case there is no address in a topic, then the event is considered a lifecycle event, and thus relevant to everyone.
+3. An event is considered relevant to the signer of the transaction that created it. 
+
+The purpose for these rules is to be simple, clear, intuitive, and to work as good as possible with the existing contracts.  
+
+In case the rules above are not providing the desired functionality, the developer can adjust the topics.
+
+For example, if one of the lifecycle events should only be visible to the administrators, the developer can add that address as a topic.
```
Comment (joeldudley) [inline design/Events_design.md:215]: Let's clarify here that you cannot create a lifecycle event strictly visible to the administrators, since the tx submitter will always be able to see it.

File: design/Events_design.md [hunk 18]
```diff
@@ -0,0 +1,300 @@
...
+
+##### Visibility Rules
+
+Users should be able to request and read all events that are relevant to them.
+
+The rules we propose are:
+
+1. An event is considered relevant to all account owners whose addresses are used as topics in the event.
+2. In case there is no address in a topic, then the event is considered a lifecycle event, and thus relevant to everyone.
+3. An event is considered relevant to the signer of the transaction that created it.
```
Comment (joeldudley) [inline design/Events_design.md:209]: Can we comment on the rationale for this. I assume it's because it's too disruptive to existing apps to hide their events from the submitter - is that right?

File: design/Events_design.md (hunk 17 above)
Comment (tudor-malene) [reply design/Events_design.md:215]: yes! very good point

**Comment by joeldudley** [comment]
@tudor-malene How do events interact with the reveal period? Do these visibility rules continue to apply once the underlying txs have been revealed (to make things simple)? Or do events become visible once the transaction has been revealed?

**Comment by tudor-malene** [comment]
> @tudor-malene How do events interact with the reveal period? Do these visibility rules continue to apply once the underlying txs have been revealed (to make things simple)? Or do events become visible once the transaction has been revealed?

good point. needs to be covered. Thanks Joel

**Comment by tudor-malene** [comment]
> Super clear and digestible now. Generally seems sensible to me, few thoughts:
> 
> * what's the lifespan for a subscription in the enclave, unsubscribe option + an expiry?
> * still pretty worried about perf and DoS potential, but we can test for it and try to optimise/prioritise traffic etc. Maybe we need read-only nodes to services these requests in production (normal nodes but the host doesn't give it any transactions/other work).
> * very minor information leak for the host about their users, host owner can see how many relevant events their subscribers are getting back. Tbf that's no different from it seeing how many transactions that user is submitting etc., just a measure of their activity I guess and not tied to their acc address at all

I'll create a section with open questions, so we keep them in mind during implementation

File: design/Events_design.md (hunk 15 above)
Comment (tudor-malene) [reply design/Events_design.md:83]: yeah, it's out of place here.

File: design/Events_design.md [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+
+What these events have in common is that they are not user-specific. They represent a general update of the smart contract.
+
+##### Visibility Rules
+
+Users should be able to request and read all events that are relevant to them.
+
+The rules we propose are:
+
+1. An event is considered relevant to all account owners whose addresses are used as topics in the event.
```
Comment (otherview) [inline design/Events_design.md:202]: Does this mean that if an event has field with my address I can see the event?

File: design/Events_design.md [hunk 20]
```diff
@@ -0,0 +1,314 @@
...
+The alternative is to remove the third rule, and allow the owner to view the event only if it contains its own address. 
+
+*Note: If we decide to remove the third rule, the implementation will have to calculate the event visibility when adding them to transaction receipts.*
+
+In case the rules above are not providing the desired functionality, the developer can adjust the topics.
+
+For example, if one of the lifecycle events should only be visible to the administrators, the developer can add that address as a topic.
+Note that the lifecycle event will be visible to the signer of the transaction as well (according to rule 3).
+
+In case an event with an address field that should not contribute to relevancy, the developer can remove the `"indexed"` and thus the event will become invisible to that user.
```
Comment (otherview) [inline design/Events_design.md:218]: Can this be expanded. In this scenario what does contribute to relevancy mean?

File: design/Events_design.md [hunk 21]
```diff
@@ -0,0 +1,314 @@
...
+There are a couple of cases that must be considered in order to decide whether Alice is entitled to view an event:
+
+1. The event was emitted by a smart contract as a result of executing a Tx sent by Alice.
+2. The event that is relevant to Alice was emitted as a result of a Tx sent by Bob. (See below for a definition of
+   relevancy.)
+3. The event that is not relevant to Alice was emitted as a result of a Tx sent by Bob.
+
+The developer of a smart contract is responsible with making sure to not include any data in the event that must be kept secret from
+the transaction originator.
+(The events are also included in the transaction receipt, which is available to the transaction submitter.)
```
Comment (otherview) [inline design/Events_design.md:82]: Are the events in the tx receipt visible to the requesting account ?

File: design/Events_design.md (hunk 19 above)
Comment (tudor-malene) [reply design/Events_design.md:202]: yes

File: design/Events_design.md (hunk 20 above)
Comment (tudor-malene) [reply design/Events_design.md:218]: being an indexed field (topic) in that log entry ( event)

File: design/Events_design.md [hunk 22]
```diff
@@ -0,0 +1,300 @@
...
+It covers two aspects:
+
+- the visibility rules for events.
+- technical implementation details.
+
+## Ethereum Events
+
+To help dApp developers design applications with a good UX, the ethereum developers invented the concept of "events" or "logs", which
+are pieces of information emitted from smart contracts, which can be streamed in real time to external applications that
+subscribe to them.
```
Comment (CaisManai) [inline design/Events_design.md:16]: I think we also need to think in terms of events as actually being written to log files. The reason we need to think this way is I'm sure tooling such as The Graph depends on reading the log files vs live streams

File: design/Events_design.md [hunk 23]
```diff
@@ -0,0 +1,300 @@
...
+(The events are also included in the transaction receipt, which is available to the transaction submitter.)
+
+#### Event relevancy
+
+In Obscuro (inherited from Ethereum), end users can have multiple accounts. The account address is how accounts are
+referenced.
+
+Events are structured objects containing multiple entries (topics or data fields).
+
+If we were designing events from scratch, with privacy in mind, we could add metadata to declare which address should be
```
Comment (CaisManai) [inline design/Events_design.md:91]: I think James's suggestion for solving problems such as Oracle requests was that we should include metadata that says this event is intended for Oracle B and then only Oracle B can query for it

File: design/Events_design.md [hunk 24]
```diff
@@ -0,0 +1,300 @@
...
+
+What these events have in common is that they are not user-specific. They represent a general update of the smart contract.
+
+##### Visibility Rules
+
+Users should be able to request and read all events that are relevant to them.
+
+The rules we propose are:
+
+1. An event is considered relevant to all account owners whose addresses are used as topics in the event.
```
Comment (CaisManai) [inline design/Events_design.md:202]: Can we discuss this particular point on call - what do you mean by 'topics in the event'. If I'm being sent some USDT, can we determine that I am a topic?
